projects/igniteui-angular/src/lib/grids/grid-base.directive.ts

```diff
@@ -3990,6 +4022,88 @@ export class IgxGridBaseDirective extends DisplayDensityBase implements
         return col.unpin(index);
     }
 
+    /**
+     * Pin the row by its id.
+     * @remarks
+     * ID is either the primaryKey value or the data record instance.
+     * @example
+     * ```typescript
+     * this.grid.pinRow(rowID);
+     * ```
+     * @param rowID The row id - primaryKey value or the data record instance.
+     * @param index The index at which to insert the row in the pinned collection.
+     */
+    public pinRow(rowID, index?): boolean {
+        const rec = this.gridAPI.get_rec_by_id(rowID);
+        if (this.pinnedRecords.indexOf(rec) !== -1 || !rec || this.data.indexOf(rec) === -1) {
```

Comment:
What's the point behind the `this.data.indexOf(rec) === -1` check? The `gridAPI.get_rec_by_id()` call should've already checked that and it handles transactions as well.
`if (!rec || this.pinnedRecords.indexOf(rec) !== -1)` would make more sense as a check.

Reply:
@damyanpetev The rowID can be either the primary key value (if primaryKey exists) or the record instance itself.
In case it is a record  get_rec_by_id does not do any additional checks to see if the passed record exists in the data or not - it simply returns the record instance itself. Hence the additional check.

At this time we don't want to allow pinning data that does not exist in the grid. We will mark this as a limitation for scenario where the user may want to pin something that is not in the data (like in remote virtualization scenarios for example).